README.md
# node-portscanner

[![JavaScript Style Guide](https://img.shields.io/badge/code%20style-standard-brightgreen.svg)](http://standardjs.com/)

The portscanner module is
an asynchronous JavaScript port scanner for Node.js.

Portscanner can check a port,
or range of ports,
for 'open' or 'closed' statuses.

[Looking for maintainer](https://github.com/baalexander/node-portscanner/issues/25)!

## Install

```bash
npm install portscanner
```

## Usage

A brief example:

```javascript
var portscanner = require('portscanner')

// Checks the status of a single port
portscanner.checkPortStatus(3000, '127.0.0.1', function(error, status) {
  // Status is 'open' if currently in use or 'closed' if available
  console.log(status)
})

// Find the first available port. Asynchronously checks, so first port
// determined as available is returned.
portscanner.findAPortNotInUse(3000, 3010, '127.0.0.1', function(error, port) {
  console.log('AVAILABLE PORT AT: ' + port)
})

// Find the first port in use or blocked. Asynchronously checks, so first port
// to respond is returned.
portscanner.findAPortInUse(3000, 3010, '127.0.0.1', function(error, port) {
  console.log('PORT IN USE AT: ' + port)
})

// You can also pass array of ports to check
portscanner.findAPortInUse([3000, 3005, 3006], '127.0.0.1', function(error, port) {
  console.log('PORT IN USE AT: ' + port)
})

// And skip host param. Default is '127.0.0.1'
portscanner.findAPortNotInUse(3000, 4000, function(error, port) {
  console.log('PORT IN USE AT: ' + port)
})
```

The example directory contains a more detailed example.

## Test

```sh
npm test
```

## Future

Please create issues or pull requests
for port scanning related features
you'd like to see included.

## License (MIT)

[MIT](LICENSE)



CHANGELOG.md
# 2.0.0

## Breaking Changes

* Ports must be numbers. Some of the new changes are based on the assumption that ports must be numbers.

## New

* [[`84db394996`](https://github.com/baalexander/node-portscanner/commit/84db394996)] - promise support (laggingreflex)
* [[`3b90e2ca74`](https://github.com/baalexander/node-portscanner/commit/3b90e2ca74)] - Improve arguments parsing (laggingreflex)
* [[`eb345da68c`](https://github.com/baalexander/node-portscanner/commit/eb345da68c)] - Make host parameter optional (Maciej Dudzinski)
* [[`2ed556b159`](https://github.com/baalexander/node-portscanner/commit/2ed556b159)] [[`99483dc28e`](https://github.com/baalexander/node-portscanner/commit/99483dc28e)] [#26](https://github.com/baalexander/node-portscanner/issues/26) - Add support for checking array of ports instead of range (Maciej Dudzinski)

## Fixes

* [[`da8ff250bd`](https://github.com/baalexander/node-portscanner/commit/da8ff250bd)] [#23](https://github.com/baalexander/node-portscanner/issues/23) - handle ports range provided in reverse order (laggingreflex)
* [[`4c7a88436f`](https://github.com/baalexander/node-portscanner/commit/4c7a88436f)] - Update async module for use this module in strict mode (Luca Pau)
* [[`04fea5dd4d`](https://github.com/baalexander/node-portscanner/commit/04fea5dd4d)] - Preserve error message from socket's error event (Jan Melcher)

## Misc

* [[`0beeca8cbd`](https://github.com/baalexander/node-portscanner/commit/0beeca8cbd)] - Implement JavaScript Standard Style Guide (laggingreflex)
* [[`da1ce58319`](https://github.com/baalexander/node-portscanner/commit/da1ce58319)] - more tests for different argument signatures (laggingreflex)
* [[`9fdbdc38d7`](https://github.com/baalexander/node-portscanner/commit/9fdbdc38d7)] - Add basic unit tests with Ava runner (Maciej Dudzinski)

# 1.2.0

Some recent changes in 1.1.0 [[`3b90e2ca74`](https://github.com/baalexander/node-portscanner/commit/3b90e2ca74)] [[`eb345da68c`](https://github.com/baalexander/node-portscanner/commit/eb345da68c)] were based on the assumption that the ports would always be of the type numbers. This broke portscanner in some cases [#49](https://github.com/baalexander/node-portscanner/issues/49) [#50](https://github.com/baalexander/node-portscanner/issues/50). All changes were reverted back to 1.0.0 and only some very critical fixes [[`4c7a88436f`](https://github.com/baalexander/node-portscanner/commit/4c7a88436f)] [[`04fea5dd4d`](https://github.com/baalexander/node-portscanner/commit/04fea5dd4d)] were applied to this version.


## Fixes

* [[`4c7a88436f`](https://github.com/baalexander/node-portscanner/commit/4c7a88436f)] - Update async module for use this module in strict mode (Luca Pau)
* [[`04fea5dd4d`](https://github.com/baalexander/node-portscanner/commit/04fea5dd4d)] - Preserve error message from socket's error event (Jan Melcher)

## Misc

* [[`69c648c740`](https://github.com/baalexander/node-portscanner/commit/69c648c740)] - tests - ports as strings (laggingreflex)
* [[`9fdbdc38d7`](https://github.com/baalexander/node-portscanner/commit/9fdbdc38d7)] - Add basic unit tests with Ava runner (Maciej Dudzinski)


# 1.1.1

Reverted to 1.0.0

# 1.1.0

Some breaking changes were introduced in this version which have since been reverted in 1.1.1.

Please update to either 1.1.1 or 2.0.0.

# 1.0.0

* [[`2463e64a0a`](https://github.com/baalexander/node-portscanner/commit/2463e64a0a)] - 1.0.0 (Sean Massa)
* [[`21aa98b632`](https://github.com/baalexander/node-portscanner/commit/21aa98b632)] - update readme (Sean Massa)
* [[`5fe4ab4a69`](https://github.com/baalexander/node-portscanner/commit/5fe4ab4a69)] - Merge pull request #21 from jdwilliams15/master (Sean Massa)
* [[`725afef7b4`](https://github.com/baalexander/node-portscanner/commit/725afef7b4)] - fix indent (jdwilliams15)
* [[`053b56e455`](https://github.com/baalexander/node-portscanner/commit/053b56e455)] - fixed indentation (jdwilliams15)
* [[`b1dd496633`](https://github.com/baalexander/node-portscanner/commit/b1dd496633)] - Changed socket error handler to handle 'ECONNREFUSED'. In event of ECONNREFUSED the port is available (jdwilliams15)
* [[`512cfdbf78`](https://github.com/baalexander/node-portscanner/commit/512cfdbf78)] - 0.2.3 (Sean Massa)
* [[`5526b8b4eb`](https://github.com/baalexander/node-portscanner/commit/5526b8b4eb)] - Merge pull request #19 from thomseddon/fix-end (Sean Massa)
* [[`a854ec6bd6`](https://github.com/baalexander/node-portscanner/commit/a854ec6bd6)] - Use socket.destroy() not socket.end() on successful connection (Thom Seddon)
* [[`c747ffa9de`](https://github.com/baalexander/node-portscanner/commit/c747ffa9de)] - 0.2.2 (Sean Massa)
* [[`4a1f8f811b`](https://github.com/baalexander/node-portscanner/commit/4a1f8f811b)] - Merge pull request #16 from baalexander/fix-port-finding (Sean Massa)
* [[`ff51ebe871`](https://github.com/baalexander/node-portscanner/commit/ff51ebe871)] - fix port reporting (Sean Massa)
* [[`e9070e85ca`](https://github.com/baalexander/node-portscanner/commit/e9070e85ca)] - 0.2.1 (Sean Massa)
* [[`809b7760ad`](https://github.com/baalexander/node-portscanner/commit/809b7760ad)] - 0.2.0 (Sean Massa)
* [[`87f35e5b87`](https://github.com/baalexander/node-portscanner/commit/87f35e5b87)] - Merge pull request #14 from baalexander/localhost-127.0.0.1 (Sean Massa)
* [[`0b72b83cab`](https://github.com/baalexander/node-portscanner/commit/0b72b83cab)] - switch out localhost for 127.0.0.1 (Sean Massa)
* [[`4407d6f701`](https://github.com/baalexander/node-portscanner/commit/4407d6f701)] - Merge pull request #13 from skilesare/master (Sean Massa)
* [[`3ed682ad7a`](https://github.com/baalexander/node-portscanner/commit/3ed682ad7a)] - Update portscanner.js (skilesare)
* [[`c1544a1bb3`](https://github.com/baalexander/node-portscanner/commit/c1544a1bb3)] - Update portscanner.js (skilesare)

test.js
import net from 'net'
import test from 'ava'
import portScanner from '.'

initialize()
checkPortStatus()
findPortInUse()
findPortNotInUse()
promise()
reverseOrder()
portsAsStrings()

function initialize () {
  test.before('Set #1 test server', t => {
    const server = net.createServer()
    server.listen(3000, '127.0.0.1', () => t.end())
  })
  test.before('Set #2 test server', t => {
    const server2 = net.createServer()
    server2.listen(2999, '127.0.0.1', () => t.end())
  })
}

function checkPortStatus () {
  test('checkPortStatus - taken', t => {
    t.plan(2)

    portScanner.checkPortStatus(3000, '127.0.0.1', (error, port) => {
      t.is(error, null)
      t.is(port, 'open')
    })
  })
  test('checkPortStatus - taken (without host)', t => {
    t.plan(2)

    portScanner.checkPortStatus(3000, (error, port) => {
      t.is(error, null)
      t.is(port, 'open')
    })
  })
  test('checkPortStatus - free', t => {
    t.plan(2)

    portScanner.checkPortStatus(3001, '127.0.0.1', (error, port) => {
      t.is(error, null)
      t.is(port, 'closed')
    })
  })
  test('checkPortStatus - free (without host)', t => {
    t.plan(2)

    portScanner.checkPortStatus(3001, (error, port) => {
      t.is(error, null)
      t.is(port, 'closed')
    })
  })
}

function findPortInUse () {
  test('findPortInUse - taken port in range', t => {
    t.plan(2)

    portScanner.findAPortInUse(3000, 3010, '127.0.0.1', (error, port) => {
      t.is(error, null)
      t.is(port, 3000)
    })
  })
  test('findPortInUse - taken port in range (without host)', t => {
    t.plan(2)

    portScanner.findAPortInUse(3000, 3010, (error, port) => {
      t.is(error, null)
      t.is(port, 3000)
    })
  })
  test('findPortInUse - taken port in range (without endPort)', t => {
    t.plan(2)

    portScanner.findAPortInUse(3000, '127.0.0.1', (error, port) => {
      t.is(error, null)
      t.is(port, 3000)
    })
  })
  test('findPortInUse - taken port in range (without endPort and host) ', t => {
    t.plan(2)

    portScanner.findAPortInUse(3000, (error, port) => {
      t.is(error, null)
      t.is(port, 3000)
    })
  })
  test('findPortInUse - taken port in range - ports as array', t => {
    t.plan(2)

    portScanner.findAPortInUse([2999, 3000, 3001], '127.0.0.1', (error, port) => {
      t.is(error, null)
      t.is(port, 2999)
    })
  })
  test('findPortInUse - taken port in range - ports as array (without host)', t => {
    t.plan(2)

    portScanner.findAPortInUse([2999, 3000, 3001], (error, port) => {
      t.is(error, null)
      t.is(port, 2999)
    })
  })
  test('findPortInUse - all ports in range free', t => {
    t.plan(2)

    portScanner.findAPortInUse(3001, 3010, '127.0.0.1', (error, port) => {
      t.is(error, null)
      t.false(port)
    })
  })
  test('findPortInUse - all ports in range free - ports as array', t => {
    t.plan(2)

    portScanner.findAPortInUse([3001, 3005, 3008], '127.0.0.1', (error, port) => {
      t.is(error, null)
      t.false(port)
    })
  })
  test('findPortInUse - all ports in range free - ports as array (without host)', t => {
    t.plan(2)

    portScanner.findAPortInUse([3001, 3005, 3008], (error, port) => {
      t.is(error, null)
      t.false(port)
    })
  })
}

function findPortNotInUse () {
  test('findAPortNotInUse - start from free port', t => {
    t.plan(2)

    portScanner.findAPortNotInUse(3001, 3010, '127.0.0.1', (error, port) => {
      t.is(error, null)
      t.is(port, 3001)
    })
  })
  test('findAPortNotInUse - start from taken port', t => {
    t.plan(2)

    portScanner.findAPortNotInUse(3000, 3010, '127.0.0.1', (error, port) => {
      t.is(error, null)
      t.is(port, 3001)
    })
  })
  test('findAPortNotInUse - all ports in range taken', t => {
    t.plan(2)

    portScanner.findAPortNotInUse(2999, 3000, '127.0.0.1', (error, port) => {
      t.is(error, null)
      t.false(port)
    })
  })
  test('findAPortNotInUse - with array as parameter', t => {
    t.plan(2)

    portScanner.findAPortNotInUse([3000, 3002, 2999], '127.0.0.1', (error, port) => {
      t.is(error, null)
      t.is(port, 3002)
    })
  })
}

function promise () {
  test('findAPortNotInUse - promise', t => {
    t.plan(1)

    portScanner.findAPortNotInUse(3000, 3010, '127.0.0.1').then(port => {
      t.is(port, 3001)
    })
  })
  test('findAPortNotInUse - promise (without host)', t => {
    t.plan(1)

    portScanner.findAPortNotInUse(3000, 3010).then(port => {
      t.is(port, 3001)
    })
  })
  test('findAPortNotInUse - promise (without host and endPort)', t => {
    t.plan(1)

    portScanner.findAPortNotInUse(3000).then(port => {
      t.is(port, 3001)
    })
  })
  test('findAPortInUse - promise', t => {
    t.plan(1)

    portScanner.findAPortInUse(3000, 3010, '127.0.0.1').then(port => {
      t.is(port, 3000)
    })
  })
  test('findAPortInUse - promise (error)', t => {
    t.plan(1)

    portScanner.findAPortInUse(3001, 3010, '127.0.0.1').catch(err => {
      t.not(err, null)
    })
  })
}

function reverseOrder () {
  test('findAPortNotInUse - ports in reverse order, lowest one being in use', t => {
    t.plan(2)

    portScanner.findAPortNotInUse(3005, 3000, '127.0.0.1', (error, port) => {
      t.is(error, null)
      t.is(port, 3001)
    })
  })
  test('findAPortNotInUse - ports in reverse order, highest one being in use', t => {
    t.plan(2)

    portScanner.findAPortNotInUse(3000, 2995, '127.0.0.1', (error, port) => {
      t.is(error, null)
      t.is(port, 2995)
    })
  })
}

function portsAsStrings () {
  test('checkPortStatus - taken (port as a string)', t => {
    t.plan(2)

    portScanner.checkPortStatus('3000', '127.0.0.1', (error, port) => {
      t.is(error, null)
      t.is(port, 'open')
    })
  })
  test('checkPortStatus - taken (without host) (port as a string)', t => {
    t.plan(2)

    portScanner.checkPortStatus('3000', (error, port) => {
      t.is(error, null)
      t.is(port, 'open')
    })
  })
  test('checkPortStatus - free (port as a string)', t => {
    t.plan(2)

    portScanner.checkPortStatus('3001', '127.0.0.1', (error, port) => {
      t.is(error, null)
      t.is(port, 'closed')
    })
  })
  test('checkPortStatus - free (without host) (port as a string)', t => {
    t.plan(2)

    portScanner.checkPortStatus('3001', (error, port) => {
      t.is(error, null)
      t.is(port, 'closed')
    })
  })
  test('findPortInUse - taken port in range (startPort as a string)', t => {
    t.plan(2)

    portScanner.findAPortInUse('2990', 3010, '127.0.0.1', (error, port) => {
      t.is(error, null)
      t.is(port, 2999)
    })
  })
  test('findPortInUse - taken port in range (endPort as a string)', t => {
    t.plan(2)

    portScanner.findAPortInUse(2990, '3010', '127.0.0.1', (error, port) => {
      t.is(error, null)
      t.is(port, 2999)
    })
  })
  test('findPortInUse - taken port in range (startPort and endPort as strings)', t => {
    t.plan(2)

    portScanner.findAPortInUse('3000', '3010', '127.0.0.1', (error, port) => {
      t.is(error, null)
      t.is(port, 3000)
    })
  })
}


lib/promise.js
var portscanner = require('./portscanner')

var promisified = exports

promisified.findAPortInUse = function() {
  return promiseConstructor(portscanner.findAPortInUse, arguments)
}

promisified.findAPortNotInUse = function() {
  return promiseConstructor(portscanner.findAPortNotInUse, arguments)
}

promisified.checkPortStatus = function() {
  return promiseConstructor(portscanner.checkPortStatus, arguments)
}

promisified.findAPortWithStatus = function() {
  return promiseConstructor(portscanner.findAPortWithStatus, arguments)
}

function promiseConstructor(method, args) {
  return new Promise(function(resolve, reject) {
    args = [].slice.call(args).concat([callback])
    method.apply(null, args)

    function callback(error, port) {
      if (error || port === false) {
        reject(error || new Error('No open port found'))
      } else {
        resolve(port)
      }
    }
  })
}


example/portscan.js
var http = require('http')
var portscanner = require('../lib/portscanner.js')

// Sets up an HTTP server listening on port 3005
var server = http.createServer(function (request, response) {
})

server.listen(3005, '127.0.0.1', function () {
  // Checks the status of an individual port.
  portscanner.checkPortStatus(3005, '127.0.0.1', function (error, status) {
    // Status should be 'open' since the HTTP server is listening on that port
    console.log('Status at port 3005 is ' + status)
    if (error) console.error(error)
  })
  portscanner.checkPortStatus(3000, '127.0.0.1', function (error, status) {
    // Status should be 'closed' since no service is listening on that port.
    console.log('Status at port 3000 is ' + status)
    if (error) console.error(error)
  })

  // Finds a port that a service is listening on
  portscanner.findAPortInUse(3000, 3010, '127.0.0.1', function (error, port) {
    // Port should be 3005 as the HTTP server is listening on that port
    console.log('Found an open port at ' + port)
    if (error) console.error(error)
  })
  // Finds a port no service is listening on
  portscanner.findAPortNotInUse(3000, 3010, '127.0.0.1', function (error, port) {
    // Will return any number between 3000 and 3010 (inclusive), that's not 3005.
    // The order is unknown as the port status checks are asynchronous.
    console.log('Found a closed port at ' + port)
    if (error) console.error(error)
  })
})



lib/portscanner.js
var net = require('net')
var dns = require('dns')
var Socket = net.Socket
var async = require('async')
var promisify = require('./promisify')

/**
 * Finds the first port with a status of 'open', implying the port is in use and
 * there is likely a service listening on it.
 */
/**
 * @param {Number} startPort - Port to begin status check on (inclusive).
 * @param {Number} [endPort=65535] - Last port to check status on (inclusive).
 * @param {String} [host='127.0.0.1'] - Host of where to scan.
 * @param {findPortCallback} [callback] - Function to call back with error or results.
 * @returns {Promise}
 * @example
 * // scans through 3000 to 3002 (inclusive)
 * portscanner.findAPortInUse(3000, 3002, '127.0.0.1', console.log)
 * // returns a promise in the absence of a callback
 * portscanner.findAPortInUse(3000, 3002, '127.0.0.1').then(console.log)
 * @example
 * // scans through 3000 to 65535 on '127.0.0.1'
 * portscanner.findAPortInUse(3000, console.log)
 */
/**
 * @param {Array} postList - Array of ports to check status on.
 * @param {String} [host='127.0.0.1'] - Host of where to scan.
 * @param {findPortCallback} [callback] - Function to call back with error or results.
 * @returns {Promise}
 * @example
 * // scans 3000 and 3002 only, not 3001.
 * portscanner.findAPortInUse([3000, 3002], console.log)
 */
function findAPortInUse () {
  var params = [].slice.call(arguments)
  params.unshift('open')
  return findAPortWithStatus.apply(null, params)
}

/**
 * Finds the first port with a status of 'closed', implying the port is not in
 * use. Accepts identical parameters as {@link findAPortInUse}
 */
function findAPortNotInUse () {
  var params = [].slice.call(arguments)
  params.unshift('closed')
  return findAPortWithStatus.apply(null, params)
}

/**
 * Checks the status of an individual port.
 */
/**
 * @param {Number} port - Port to check status on.
 * @param {String} [host='127.0.0.1'] - Host of where to scan.
 * @param {checkPortCallback} [callback] - Function to call back with error or results.
 * @returns {Promise}
 */
/**
 * @param {Number} port - Port to check status on.
 * @param {Object} [opts={}] - Options object.
 * @param {String} [opts.host='127.0.0.1'] - Host of where to scan.
 * @param {Number} [opts.timeout=400] - Connection timeout in ms.
 * @param {checkPortCallback} [callback] - Function to call back with error or results.
 * @returns {Promise}
 */
function checkPortStatus (port) {
  var args, host, opts, callback

  args = [].slice.call(arguments, 1)

  if (typeof args[0] === 'string') {
    host = args[0]
  } else if (typeof args[0] === 'object') {
    opts = args[0]
  } else if (typeof args[0] === 'function') {
    callback = args[0]
  }

  if (typeof args[1] === 'object') {
    opts = args[1]
  } else if (typeof args[1] === 'function') {
    callback = args[1]
  }

  if (typeof args[2] === 'function') {
    callback = args[2]
  }

  if (!callback) return promisify(checkPortStatus, arguments)

  opts = opts || {}

  host = host || opts.host || '127.0.0.1'

  var timeout = opts.timeout || 400
  var connectionRefused = false

  var socket = new Socket()
  var status = null
  var error = null

  // Socket connection established, port is open
  socket.on('connect', function () {
    status = 'open'
    socket.destroy()
  })

  // If no response, assume port is not listening
  socket.setTimeout(timeout)
  socket.on('timeout', function () {
    status = 'closed'
    error = new Error('Timeout (' + timeout + 'ms) occurred waiting for ' + host + ':' + port + ' to be available')
    socket.destroy()
  })

  // Assuming the port is not open if an error. May need to refine based on
  // exception
  socket.on('error', function (exception) {
    if (exception.code !== 'ECONNREFUSED') {
      error = exception
    } else {
      connectionRefused = true
    }
    status = 'closed'
  })

  // Return after the socket has closed
  socket.on('close', function (exception) {
    if (exception && !connectionRefused) { error = error || exception } else { error = null }
    callback(error, status)
  })

  console.log({ host })
  dns.lookup(host, function (err, host) {
    console.log({ err, host })
    if (err) {
      callback(error)
    } else {
      socket.connect(port, host)
    }
  })
}
/**
 * Callback for {@link checkPortStatus}
 * @callback checkPortCallback
 * @param {Error|null} error - Any error that occurred while port scanning, or null.
 * @param {String} status - Status: 'open' if the port is in use, 'closed' if the port is available.
 */

/**
 * Internal helper function used by {@link findAPortInUse} and {@link findAPortNotInUse}
 * to find a port from a range or a list with a specific status.
 */
/**
 * @param {String} status - Status to check.
 * @param {...params} params - Params as passed exactly to {@link findAPortInUse} and {@link findAPortNotInUse}.
 */
function findAPortWithStatus (status) {
  var params, startPort, endPort, portList, host, callback

  params = [].slice.call(arguments, 1)

  if (typeof params[0] === 'number') {
    startPort = params[0]
  } else if (params[0] instanceof Array) {
    portList = params[0]
  }

  if (typeof params[1] === 'number') {
    endPort = params[1]
  } else if (typeof params[1] === 'string') {
    host = params[1]
  } else if (typeof params[1] === 'function') {
    callback = params[1]
  }

  if (typeof params[2] === 'string') {
    host = params[2]
  } else if (typeof params[2] === 'function') {
    callback = params[2]
  }

  if (typeof params[3] === 'function') {
    callback = params[3]
  }

  if (!callback) return promisify(findAPortWithStatus, arguments)

  if (startPort && endPort && endPort < startPort) {
    // WARNING: endPort less than startPort. Using endPort as startPort & vice versa.
    var tempStartPort = startPort
    startPort = endPort
    endPort = tempStartPort
  }

  endPort = endPort || 65535

  var foundPort = false
  var numberOfPortsChecked = 0
  var port = portList ? portList[0] : startPort

  // Returns true if a port with matching status has been found or if checked
  // the entire range of ports
  var hasFoundPort = function () {
    return foundPort || numberOfPortsChecked === (portList ? portList.length : endPort - startPort + 1)
  }

  // Checks the status of the port
  var checkNextPort = function (callback) {
    checkPortStatus(port, host, function (error, statusOfPort) {
      numberOfPortsChecked++
      if (statusOfPort === status) {
        foundPort = true
        callback(error)
      } else {
        port = portList ? portList[numberOfPortsChecked] : port + 1
        callback(null)
      }
    })
  }

  // Check the status of each port until one with a matching status has been
  // found or the range of ports has been exhausted
  async.until(hasFoundPort, checkNextPort, function (error) {
    if (error) {
      callback(error, port)
    } else if (foundPort) {
      callback(null, port)
    } else {
      callback(null, false)
    }
  })
}
/**
 * Callback for {@link findAPortWithStatus}, and by that extension, for {@link findAPortInUse} and {@link findAPortNotInUse}.
 * @callback findPortCallback
 * @param {Error|null} error - Any error that occurred while port scanning, or null.
 * @param {Number|Boolean} port - The first open port found. Note, this is the first port that returns status as 'open', not necessarily the first open port checked. If no open port is found, the value is false.
 */

/**
 * @exports portscanner
 */

module.exports = {
  findAPortInUse: findAPortInUse,
  findAPortNotInUse: findAPortNotInUse,
  checkPortStatus: checkPortStatus
}
